File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int b = 0, e = matrix.size() - 1, m;
        while(e >= b) {
            m = (b + e) / 2;
            if(matrix[m][0] == target) {
                return true;
            }
            else if(matrix[m][0] > target) {
                e = m - 1;
            }
            else {
                b = m + 1;
            }
        }
        m = (b + e) / 2;
        cout<<m;
        b = 0;
        e = matrix[m].size() - 1;
        int m2;
        while(b <= e) {
            m2 = (b + e) / 2;
            if(matrix[m][m2] == target) {
                return true;
            }
            else if(matrix[m][m2] < target) {
                b = m2 + 1;
            }
            else {
                e = m2 - 1;
            }
        }
        return false;
    }
};

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        beg, end = 0, len(matrix) - 1
        mid = (beg + end) // 2
        while beg < end:
            if matrix[mid][0] <= target <= matrix[mid][-1]:
                break
            elif matrix[mid][0] > target:
                end = mid - 1
            else:
                beg = mid + 1
            mid = (beg + end) // 2
        row = matrix[mid]
        beg, end = 0, len(row) - 1
        mid = (beg + end) // 2
        while beg <= end:
            if row[mid] < target:
                beg = mid + 1
            elif row[mid] > target:
                end = mid - 1
            else:
                return True
            mid = (beg + end) // 2
        return False